### PR TITLE
Sigstore sign

### PR DESCRIPTION
**Pull request by mtrmac**
- Allow creating sigstore signatures via `--sign-by-sigstore-private-key` .  Like existing `--sign-by`, it does not work remote (in this case because we would have to copy the private key to the server).
- Allow passing a passphrase (which is mandatory for sigstore private keys) via `--sign-passphrase-file`; if it is not provided, prompt interactively.
- Also, use that passphrase for `--sign-by` as well, allowing non-interactive GPG use. (But `--sign-passphrase-file` can only be used with _one of_ `--sign-by` and `--sign-by-sigstore-private-key`.)

Note that unlike the existing code, `podman build` does not yet implement sigstore (I'm not sure why it needs to, it seems not to push images?) because Buildah does not expose the feature yet.

Also, `podman image sign` was not extended to support sigstore.

See individual commit messages for details.

#### Does this PR introduce a user-facing change?

```release-note
Added the ability to create sigstore signatures in `podman push` and `podman manifest push`.
Added an option to read image signing passphrase from a file.
```

### CODE REVIEW

**Comment by mtrmac** [comment]
Draft:
- Depends on unmerged https://github.com/containers/common/pull/1106
- I didn’t even compile-test this yet; I’ll try this tomorrow.

**Comment by mtrmac** [comment]
This implements the `podman push` part of #7705. (Not the `podman image sign` part.)

File: test/e2e/push_test.go [hunk 1]
```diff
@@ -105,7 +106,9 @@ var _ = Describe("Podman push", func() {
 		}
 		lock := GetPortLock("5000")
 		defer lock.Unlock()
-		session := podmanTest.Podman([]string{"run", "-d", "--name", "registry", "-p", "5000:5000", REGISTRY_IMAGE, "/entrypoint.sh", "/etc/docker/registry/config.yml"})
+		// FIXME. This uses registry-1.docker.io to bypass a registries.conf mirror configuration
+		// Actually we need to mirror that image.
+		session := podmanTest.Podman([]string{"run", "-d", "--name", "registry", "-p", "5000:5000", "registry-1.docker.io/library/registry:2.8.1", "/entrypoint.sh", "/etc/docker/registry/config.yml"})
```
Comment (mtrmac) [inline test/e2e/push_test.go:111]: The current `REGISTRY_IMAGE` is `quay.io/libpod/registry:2.6`, which does not support OCI artifacts.

So the `quay.io/libpod/registry` image should be updated, to at least 2.7. And there already seems to be a 2.8 image in that repo – so let me try updating that, in a separate PR.

**Comment by mtrmac** [comment]
This now includes (via a merge commit) #15122, so that tests can run.

After the registry update PR is merged is merged, this can be rebased on top — or, I guess, merged as is.

File: test/e2e/push_test.go (hunk 1 above)
Comment (mtrmac) [reply test/e2e/push_test.go]: #15122

**Comment by mtrmac** [comment]
Ready for review, see https://github.com/containers/podman/pull/15108#issuecomment-1200078595 about whether it should be merged.

**Review by vrothberg — COMMENTED** [review]
LGTM

**Comment by vrothberg** [comment]
> Ready for review, see [#15108 (comment)](https://github.com/containers/podman/pull/15108#issuecomment-1200078595) about whether it should be merged.

I am good to merge as is.

@rhatdan @Luap99 PTAL

File: test/e2e/testdata/sigstore-key.key [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-----BEGIN ENCRYPTED COSIGN PRIVATE KEY-----
```
Comment (Luap99) [inline test/e2e/testdata/sigstore-key.key:1]: Do we really want to commit private keys for testing? IMO keys should be automatically generated before the tests and we should not used fixed ones.

File: test/utils/utils.go [hunk 3]
```diff
@@ -225,7 +225,7 @@ func (p *PodmanTest) WaitContainerReady(id string, expStr string, timeout int, s
 			return false
 		}
 
-		if strings.Contains(s.OutputToString(), expStr) {
+		if strings.Contains(s.OutputToString(), expStr) || strings.Contains(s.ErrorToString(), expStr) {
```
Comment (flouthoc) [inline test/utils/utils.go:228]: This looks a bit weird, why are we matching stderr for a successful case. Should it be a different commit ?

Comment (flouthoc) [reply test/utils/utils.go:228]: Never mind I think reason for this is specified  in the commit [Update the registry server we test against from 2.6 to 2.8](https://github.com/containers/podman/pull/15108/commits/5a5624f8189c0e4c529e4c00d6ed66397d1db5b4) .

Comment (mtrmac) [reply test/utils/utils.go:228]: Yes, that’s the reason. `WaitContainerReady` is most commonly used for the registry, so I took this fairly lazy route.

Alternatively, we could add an option whether to check `stderr` to this function, or maybe a specialized “start a registry and wait until ready” helper.

File: test/e2e/testdata/sigstore-key.key (hunk 2 above)
Comment (mtrmac) [reply test/e2e/testdata/sigstore-key.key:1]: The format for this is `cosign`-specific. Right now, to generate this, we would need to install `cosign` during tests.

Eventually Skopeo will be taught to generate the key pair, and/or we might accept keys in more standardized formats, but that doesn’t exist yet.

(To be explicit, this key pair was generated specifically for the tests and has no other use.)

Comment (Luap99) [reply test/e2e/testdata/sigstore-key.key:1]: I am fine with this, looks like we already have other private keys committed anyway.

My main concern is that we are testing this *one* key and not random generated keys like normal users would do.

**Review by Luap99 — COMMENTED** [review]
LGTM
